src/pages/ProvideLiquidity/AddLiquidityForm.tsx

```diff
@@ -0,0 +1,180 @@
+  const theme = useContext(ThemeContext)
+
+  const { network } = useContractKit()
+
+  const { independentField, typedValue, otherTypedValue } = useMintState()
+  const {
+    currencies,
+    price,
+    noLiquidity,
+    poolTokenPercentage,
+    dependentField,
+    parsedAmounts,
+    currencyBalances,
+    error,
+  } = useDerivedMintInfo(token0, token1)
+
+  const { onFieldAInput, onFieldBInput } = useMintActionHandlers(noLiquidity)
+
+  // check whether the user has approved the router on the tokens
+  const [approvalA, approveACallback] = useApproveCallback(parsedAmounts[Field.CURRENCY_A], ROUTER_ADDRESS)
+  const [approvalB, approveBCallback] = useApproveCallback(parsedAmounts[Field.CURRENCY_B], ROUTER_ADDRESS)
+
+  const chainId = network.chainId
+
+  // get formatted amounts
+  const formattedAmounts = {
+    [independentField]: typedValue,
+    [dependentField]: noLiquidity ? otherTypedValue : parsedAmounts[dependentField]?.toSignificant(6) ?? '',
+  }
+
+  // get the max amounts user can add
+  const maxAmounts: { [field in Field]?: TokenAmount } = [Field.CURRENCY_A, Field.CURRENCY_B].reduce(
+    (accumulator, field) => {
+      return {
+        ...accumulator,
+        [field]: maxAmountSpend(currencyBalances[field]),
+      }
+    },
+    {}
+  )
+
+  const atMaxAmounts: { [field in Field]?: TokenAmount } = [Field.CURRENCY_A, Field.CURRENCY_B].reduce(
+    (accumulator, field) => {
+      return {
+        ...accumulator,
+        [field]: maxAmounts[field]?.equalTo(parsedAmounts[field] ?? '0'),
+      }
+    },
+    {}
+  )
+
+  return (
+    <>
+      <CurrencyInputPanel
+        value={formattedAmounts[Field.CURRENCY_A]}
+        onUserInput={onFieldAInput}
+        onMax={() => {
+          if (currencies[Field.CURRENCY_A]?.address === CELO[chainId as unknown as UbeswapChainId].address) {
+            onFieldAInput(Math.max(Number(maxAmounts[Field.CURRENCY_A]?.toExact() ?? '') - 0.01, 0).toString())
+          } else {
+            onFieldAInput(maxAmounts[Field.CURRENCY_A]?.toExact() ?? '')
+          }
+        }}
+        showMaxButton={!atMaxAmounts[Field.CURRENCY_A]}
+        currency={token0}
+        id="add-liquidity-input-token-a"
+        disableCurrencySelect
+      />
+      <StyledColumnCenter>
+        <Plus size="16" color={theme.text2} />
+      </StyledColumnCenter>
+      <CurrencyInputPanel
+        value={formattedAmounts[Field.CURRENCY_B]}
+        onUserInput={onFieldBInput}
+        onMax={() => {
+          if (currencies[Field.CURRENCY_B]?.address === CELO[chainId as unknown as UbeswapChainId].address) {
+            onFieldBInput(Math.max(Number(maxAmounts[Field.CURRENCY_B]?.toExact() ?? '') - 0.01, 0).toString())
+          } else {
+            onFieldBInput(maxAmounts[Field.CURRENCY_B]?.toExact() ?? '')
+          }
+        }}
+        showMaxButton={!atMaxAmounts[Field.CURRENCY_B]}
+        currency={token1}
+        id="add-liquidity-input-token-b"
+        disableCurrencySelect
+      />
```

Comment:
nit: this could be more DRY

Reply:
do you think so?? the 2 inputs look similar but all the props are different (token0 vs token1) so i don't think abstracting it one more layer helps readability...

Reply:
yeah I think so. Reading this to look for bugs, I need to look at 9 different places to verify that the correct token was used (1 or 2) and 4 different lines of logic to verify that it's the same in both places. If it were in a helper function, I'd only need to look at 3-4 parameters, and the logic would be guaranteed the same (since it's re-used). That's 9-10 fewer things to think about. 

But whatever, it's a nit.